lib/utils/nd_list.dart

```diff
@@ -0,0 +1,629 @@
+    if (priorResult.shape.isEmpty) {
+      throw ArgumentError('Cannot index an empty NDList');
+    }
+    while (index < 0) {
+      // -1 => priorResult.shape[0] - 1 (aka last element)
+      // -2 => second last element, etc.
+      index += priorResult.shape[0];
+    }
+    // error handling
+    if (index >= priorResult.shape[0]) {
+      throw RangeError(
+          'Index out of bounds: index $index is out of bounds for axis with size ${priorResult.shape[0]}');
+    }
+    // return the appropriate axis-0 slice
+    if (priorResult.shape.length == 1) {
+      return priorResult.resolveStep([index], [1]);
+    }
+    final returnShape = priorResult.shape.sublist(1);
+    final subLength = _product(returnShape);
+    final listIndex = List.generate(subLength, (i) => index * subLength + i);
+    return priorResult.resolveStep(listIndex, returnShape);
+  }
+
+  /// This builds on the base case of an axis-0 int index, and allows for indexing on any axis.
+  static NDIndexResult<X> _intIndexWithAxis<X>(
+      NDIndexResult<X> priorResult, int index, int axis) {
+    return _slice(priorResult, index, index + 1, axis: axis);
+  }
+
+  NDList<X> slice(int start, int end, {int axis = 0}) {
+    return _slice(NDIndexResult.from(this), start, end, axis: axis).evaluate();
+  }
+
+  static NDIndexResult<Y> _slice<Y>(
+      NDIndexResult<Y> priorResult, int start, int end,
+      {int axis = 0}) {
+    // if (start < 0) {
+    //   start += priorResult.shape[axis];
+    // }
+    // if (end < 0) {
+    //   end += priorResult.shape[axis];
+    // }
+    // if (end < start) {
+    //   return _slice(priorResult, end, start, axis: axis);
+    // }
+    if (end == start) {
+      return priorResult.resolveStep([], []);
+    }
+    if (priorResult.shape.length == 1) {
+      final sliceEnd = end > priorResult.shape[0] ? priorResult.shape[0] : end;
+      final sliceLength = sliceEnd - start;
+      final indices = List.generate(sliceLength, (i) => start + i);
+
+      return priorResult.resolveStep(indices, [sliceLength]);
+    }
+
+    if (axis > priorResult.shape.length - 1) {
+      throw ArgumentError(
+          'Invalid axis $axis for ${priorResult.shape.length}D list with shape $priorResult.shape');
+    }
+
+    if (axis == 0) {
+      // final singleSliceShape = priorResult.shape.sublist(1);
+      // final sliceStep = _product(singleSliceShape);
+      final sliceEnd =
+          end; // > priorResult.shape[0] ? priorResult.shape[0] : end;
+      final sliceLength = sliceEnd - start;
+      final listIndices = [
+        for (int i = start; i < sliceEnd; i++)
+          ..._intIndex(priorResult, i).parentIndices
+      ];
+
+      final sliceShape = [sliceLength, ...priorResult.shape.sublist(1)];
+      return NDIndexResult(priorResult.parent, listIndices, sliceShape);
+    }
+
+    final subtensorIndices = _enumerateSubtensors(priorResult.shape, axis);
+    final indicesAndSubTensors = subtensorIndices
```

Comment:
I've been trying to debug this test:
https://github.com/ericcanton/sleep_classifier_app/blob/19cff016fcf102153a1e0ae716aef9cad022db7e/test/utils_tests/nd_list_test.dart#L110-L131

that wasn't passing on my end. The issue is definitely here, when you get the subtensors for `axis=1`. I'm still confused about the recursive call to `_slice` and why `axis: axis - 1` on line 421